CentralHub.Api/Services/LocalizationService.cs

```diff
@@ -4,20 +4,63 @@ namespace CentralHub.Api.Services;
 
 public class LocalizationService : ILocalizationService
 {
-    private Dictionary<int, List<Measurement>> _measurements = new Dictionary<int, List<Measurement>>();
-
+    private Thread? MeasurementRemover;
+    private readonly List<ThreadState> validStates = new List<ThreadState>() { ThreadState.Running, ThreadState.WaitSleepJoin };
+    private Dictionary<int, List<MeasurementGroup>> _measurements = new Dictionary<int, List<MeasurementGroup>>();
     public void AddMeasurements(int id, IReadOnlyCollection<Measurement> measurements)
     {
         lock (_measurements)
         {
-            if (_measurements.TryGetValue(id, out List<Measurement> value))
+            if (_measurements.TryGetValue(id, out List<MeasurementGroup> value))
             {
-                value.AddRange(measurements);
+                value.Add(new MeasurementGroup(measurements.ToList()));
             }
             else
             {
-                _measurements.Add(id, measurements.ToList());
+                _measurements.Add(id, new List<MeasurementGroup>() { new MeasurementGroup(measurements.ToList()) });
+            }
+        }
+        if (MeasurementRemover == null || !validStates.Contains(MeasurementRemover.ThreadState))
+        {
+            MeasurementRemover = new Thread(new ThreadStart(RemoveMeasurements));
+            MeasurementRemover.Start();
+        }
+    }
+    private void RemoveMeasurements()
+    {
+        var toBeRemoved = new List<MeasurementGroup>();
+        while (true)
+        {
+            var delay = 120;
+            lock (_measurements)
+            {
+                foreach (var key in _measurements.Keys)
+                {
+                    toBeRemoved.Clear();
+                    foreach (var group in _measurements[key])
+                    {
+                        if (DateTime.Now >= group.Timestamp + TimeSpan.FromMinutes(2))
+                        {
+                            toBeRemoved.Add(group);
+                        }
+                        else
+                        {
+                            var new_delay = (group.Timestamp + TimeSpan.FromMinutes(2) - DateTime.Now).TotalSeconds;
+                            delay = new_delay < delay ? (int)new_delay : delay;
+
+                        }
+                    }
+                    foreach (var group in toBeRemoved)
+                    {
+                        _measurements[key].Remove(group);
+                    }
+                }
+                if (_measurements.Keys.All(key => _measurements[key].Count == 0))
+                {
+                    return;
+                }
             }
+            Thread.Sleep(delay * 1000);
```

Comment:
Sleep until the next group needs to be removed

Reply:
that is what it does? it calculates delay during the thread's execution

Reply:
i should update it to use milliseconds in the calculation instead of multiplying by 1000 tho

Reply:
Ok I see it just seems strange to me that delay is initialized to 120

Reply:
Max value

Reply:
Set it to int.MaxValue

Reply:
There is no need to wakeup to do nothing